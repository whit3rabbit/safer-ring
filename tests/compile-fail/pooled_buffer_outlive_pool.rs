// This test should fail to compile because pooled buffers cannot outlive their pool

use safer_ring::{BufferPool, Ring};

fn main() {
    let ring = Ring::new(10).unwrap();
    let mut buffer = {
        let pool = BufferPool::new(10, 1024);
        pool.get().unwrap() // This buffer should not be able to outlive the pool
    };
    
    // This should not compile - buffer references dropped pool
    let _future = ring.read(0, buffer.as_mut_slice());
}