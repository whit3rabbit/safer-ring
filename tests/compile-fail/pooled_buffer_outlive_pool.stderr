error[E0596]: cannot borrow `ring` as mutable, as it is not declared as mutable
  --> tests/compile-fail/pooled_buffer_outlive_pool.rs:13:19
   |
13 |     let _future = ring.read(0, buffer.as_mut_slice());
   |                   ^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
6  |     let mut ring = Ring::new(10).unwrap();
   |         +++

error[E0597]: `buffer` does not live long enough
  --> tests/compile-fail/pooled_buffer_outlive_pool.rs:13:32
   |
7  |     let mut buffer = {
   |         ---------- binding `buffer` declared here
...
13 |     let _future = ring.read(0, buffer.as_mut_slice());
   |                                ^^^^^^ borrowed value does not live long enough
14 | }
   | -
   | |
   | `buffer` dropped here while still borrowed
   | borrow might be used here, when `ring` is dropped and runs the `Drop` code for type `Ring`
   |
   = note: values in a scope are dropped in the opposite order they are defined
